pkg/streamingpromql/operators/instant_vector_selector_test.go

```diff
@@ -0,0 +1,173 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package operators
+
+import (
+	"context"
+	"testing"
+	"time"
+
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/model/timestamp"
+	"github.com/prometheus/prometheus/promql"
+	"github.com/prometheus/prometheus/promql/promqltest"
+	"github.com/stretchr/testify/require"
+
+	"github.com/grafana/mimir/pkg/streamingpromql/pooling"
+)
+
+func TestInstantVectorSelector_NativeHistogramPointerHandling(t *testing.T) {
+	requireNotSame := func(t *testing.T, h1, h2 *histogram.FloatHistogram) {
+		require.NotSame(t, h1, h2, "must not point to the same *FloatHistogram")
+		require.NotSame(t, h1.PositiveSpans, h2.PositiveSpans, "must not point to the same positive spans slice")
+		require.NotSame(t, h1.NegativeSpans, h2.NegativeSpans, "must not point to the same negative spans slice")
+		require.NotSame(t, h1.PositiveBuckets, h2.PositiveBuckets, "must not point to the same positive buckets slice")
+		require.NotSame(t, h1.NegativeBuckets, h2.NegativeBuckets, "must not point to the same negative buckets slice")
+		require.NotSame(t, h1.CustomValues, h2.CustomValues, "must not point to the same custom values slice")
+	}
+
+	testCases := map[string]struct {
+		data      string
+		stepCount int
+		check     func(t *testing.T, points []promql.HPoint, floats []promql.FPoint)
+	}{
+		"different histograms at each point": {
+			data: `
+                load 1m
+                    my_metric {{schema:0 sum:5 count:4 buckets:[1 2 1]}} {{schema:0 sum:20 count:7 buckets:[9 10 1]}} {{schema:0 sum:21 count:8 buckets:[9 10 2]}}
+                #             0m										 1m                                           2m
+            `,
+			stepCount: 3,
+			check: func(t *testing.T, points []promql.HPoint, _ []promql.FPoint) {
+				require.Len(t, points, 3)
+				requireNotSame(t, points[0].H, points[1].H)
+				requireNotSame(t, points[1].H, points[2].H)
```

Comment:
Also, it's possible that points[2] is the same as points[0], and both are different from points[1]

Reply:
Fixed in 4ea8165